crates/rwasm/machine/src/alu/extend/mod.rs

```diff
@@ -0,0 +1,336 @@
+        cols.is_extend16s = F::from_bool(event.opcode == Opcode::I32Extend16S);
+
+        let b = event.b;
+        let b_bytes = b.to_le_bytes();
+
+        // Emit one lookup per active opcode and set the unified `msb` column.
+        if event.opcode == Opcode::I32Extend8S {
+            let msb = (b_bytes[0] >> 7) & 1;
+            cols.msb = F::from_canonical_u32(msb as u32);
+            blu.add_byte_lookup_event(ByteLookupEvent::new(
+                ByteOpcode::MSB,
+                msb as u16,
+                0,
+                b_bytes[0],
+                0,
+            ));
+        } else if event.opcode == Opcode::I32Extend16S {
+            let msb = (b_bytes[1] >> 7) & 1;
+            cols.msb = F::from_canonical_u32(msb as u32);
+            blu.add_byte_lookup_event(ByteLookupEvent::new(
+                ByteOpcode::MSB,
+                msb as u16,
+                0,
+                b_bytes[1],
+                0,
+            ));
+        }
+    }
+}
+
+impl<AB> Air<AB> for ExtendChip
+where
+    AB: SP1CoreAirBuilder,
+{
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local = main.row_slice(0);
+        let local: &ExtendCols<AB::Var> = (*local).borrow();
+
+        builder.assert_bool(local.is_extend8s);
+        builder.assert_bool(local.is_extend16s);
+        builder.assert_bool(local.msb);
+
+        let is_real = local.is_extend8s + local.is_extend16s;
+        builder.assert_bool(is_real.clone());
+
+        let ff = AB::Expr::from_canonical_u32(0xFF);
+        let a = &local.a.0;
+        let b = &local.b.0;
+
+        // Constrain the unified `msb` column depending on the active opcode.
+        // The correct MSB is certified by the corresponding lookup.
+        builder.send_byte(
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.msb,         // Certified result
+            b[0],              // op1: byte being checked
+            AB::Expr::zero(),  // op2
+            local.is_extend8s, // Multiplicity
+        );
+        builder.send_byte(
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.msb,          // Certified result
+            b[1],               // op1: byte being checked
+            AB::Expr::zero(),   // op2
+            local.is_extend16s, // Multiplicity
+        );
+
+        // i32.extend8_s result bytes
+        builder.when(local.is_extend8s).assert_zero(a[0] - b[0]);
+        builder.when(local.is_extend8s).assert_zero(a[1] - local.msb * ff.clone());
+        builder.when(local.is_extend8s).assert_zero(a[2] - local.msb * ff.clone());
+        builder.when(local.is_extend8s).assert_zero(a[3] - local.msb * ff.clone());
+
+        // i32.extend16_s result bytes
+        builder.when(local.is_extend16s).assert_zero(a[0] - b[0]);
+        builder.when(local.is_extend16s).assert_zero(a[1] - b[1]);
+        builder.when(local.is_extend16s).assert_zero(a[2] - local.msb * ff.clone());
```

Comment:
for a2 a3 you can use builder.when(local.msb).assert_eq(a[2], ff.clone()) builder.when(local.msb).assert_eq(a[3], ff.clone())